syntax = "proto3";

import "google/protobuf/empty.proto";

package paxos;

/*
 * Paxos service definition
 */
service Paxos {

    // Function to request transfer of amount
    rpc Transfer (TransferReq) returns (TransferRes) {}

    // Function to print client's balance
    rpc GetBalance (google.protobuf.Empty) returns (Balance) {}

    // Function to print server's local logs
    rpc GetLogs (google.protobuf.Empty) returns (Logs) {}

    // Function to print server's persisted transaction blocks
    rpc GetDBLogs (google.protobuf.Empty) returns (Logs) {}

    // Function to send prepare request
    rpc Prepare (PrepareReq) returns (PrepareRes) {}

    // Function to request acceptance of a value
    rpc Accept (AcceptReq) returns (AcceptRes) {}

    // Function to commit a transaction block
    rpc Commit (CommitReq) returns (CommitRes) {}

    // Function to replicate transaction block
    rpc Success (SuccessReq) returns (SuccessRes) {}
}


/*
 * Message definitions
 */

// Money transfer request
message TransferReq {
    string receiver = 1;
    int32 amount = 2;
}

// Response to money transfer request
message TransferRes {
    bool ack = 1;
    string server_id = 2;
}

message Balance {
    int32 amount = 1;
}

// Proposal that is sent during the prepare phase
message Proposal {
    int32 number = 1;
    string server_id = 2;
}

// An individual transaction 
message Transaction {
    int32 id = 1;
    string sender = 2;
    string receiver = 3;
    int32 amount = 4;
}

message Logs {
    repeated Transaction logs = 1;
}

// The prepare message to initiate consensus
message PrepareReq {
    Proposal proposal = 1;
    int32 last_committed_block = 2;
}

// The promise message sent as response to prepare
message PrepareRes {
    bool ack = 1;
    Proposal proposal = 2;
    optional Proposal accept_num = 3;
    optional Logs accept_val = 4;
    optional Logs logs = 5;
}

// Accept request sent by proposer
message AcceptReq {
    Proposal proposal = 1;
    int32 last_committed_block = 2;
    Logs logs = 3;
}

// Accept response sent by acceptors
message AcceptRes {
    bool ack = 1;
    Proposal proposal = 2;
    string server_id = 3;
    optional int32 last_committed_block = 4;    // needed for server to decide whether to catch-up or not
}

// Commit request sent by proposer
message CommitReq {
    Proposal proposal = 1;
    int32 last_committed_block = 2;
}

// Commit response sent by acceptor
message CommitRes {
    bool ack = 1;
    Proposal proposal = 2;
}

// Combination of AcceptReq + CommitReq (used for catch up)
message SuccessReq {
    // proposal number corresponding to the committed block
    Proposal proposal = 1;
    int32 block_id = 2;
    Logs block = 4;
}

message SuccessRes {
    bool ack = 1;
    string server_id = 2;
    int32 last_committed_block = 3;
}